File: benchmarks/lcs/nim/lcs.nim
import std/strutils
import std/os

# splitmix64 (public domain, Vigna). A single multiply-mod step (as used in
# the quicksort benchmark) turned out to have too little avalanche once
# reduced to a 4-symbol alphabet (near-identical buffers, LCS almost n);
# splitmix64's xor-shift/multiply finalizer avalanches properly even after
# `mod 4`.
proc splitmix64(xIn: uint64): uint64 =
    var x = xIn
    x = x + 0x9E3779B97F4A7C15'u64
    x = (x xor (x shr 30)) * 0xBF58476D1CE4E5B9'u64
    x = (x xor (x shr 27)) * 0x94D049BB133111EB'u64
    x = x xor (x shr 31)
    return x

type ByteArr = ptr UncheckedArray[uint8]
type IntArr = ptr UncheckedArray[int32]

# Different seed per buffer (not just a shifted index into the same stream)
# so A and B are independent-looking sequences, not one shifted copy of the
# other.
proc fill(buf: ByteArr, n: int32, seed: uint64) =
    for i in 0 ..< n:
        let idx = uint64(i + 1) + seed
        let h = splitmix64(idx) mod 4
        buf[i] = uint8(65 + h.int)

proc maxI32(a, b: int32): int32 =
    if a > b: a else: b

proc lcsLength(a, b: ByteArr, n: int32): int32 =
    var prev = cast[IntArr](alloc((n + 1) * sizeof(int32)))
    var curr = cast[IntArr](alloc((n + 1) * sizeof(int32)))
    for j in 0 .. n:
        prev[j] = 0

    for i in 1 .. n:
        curr[0] = 0
        for j in 1 .. n:
            if a[i - 1] == b[j - 1]:
                curr[j] = prev[j - 1] + 1
            else:
                curr[j] = maxI32(prev[j], curr[j - 1])
        swap(prev, curr)

    result = prev[n]
    dealloc(prev)
    dealloc(curr)

when isMainModule:
    if paramCount() < 1:
        stderr.writeLine("Usage: lcs <n>")
        quit(1)
    let n = parseInt(paramStr(1)).int32

    if n <= 0:
        echo 0
        quit(0)

    let a = cast[ByteArr](alloc(int(n)))
    let b = cast[ByteArr](alloc(int(n)))
    fill(a, n, 0'u64)
    fill(b, n, 0x9E3779B9'u64)

    let result = lcsLength(a, b, n)
    dealloc(a)
    dealloc(b)
    echo result

File: benchmarks/lcs/zig/lcs.zig
const std = @import("std");

// splitmix64 (public domain, Vigna). A single multiply-mod step (as used in
// the quicksort benchmark) turned out to have too little avalanche once
// reduced to a 4-symbol alphabet (near-identical buffers, LCS almost n);
// splitmix64's xor-shift/multiply finalizer avalanches properly even after
// `% 4`.
fn splitmix64(x_in: u64) u64 {
    var x = x_in;
    x +%= 0x9E3779B97F4A7C15;
    x = (x ^ (x >> 30)) *% 0xBF58476D1CE4E5B9;
    x = (x ^ (x >> 27)) *% 0x94D049BB133111EB;
    x = x ^ (x >> 31);
    return x;
}

// Different seed per buffer (not just a shifted index into the same stream)
// so A and B are independent-looking sequences, not one shifted copy of the
// other.
fn fill(buf: []u8, n: i32, seed: u64) void {
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        const idx: u64 = @as(u64, @intCast(i + 1)) +% seed;
        const h = splitmix64(idx) % 4;
        buf[@intCast(i)] = @intCast(@as(u64, 'A') + h);
    }
}

fn maxI32(a: i32, b: i32) i32 {
    return if (a > b) a else b;
}

fn lcsLength(allocator: std.mem.Allocator, a: []const u8, b: []const u8, n: i32) !i32 {
    var prev = try allocator.alloc(i32, @intCast(n + 1));
    defer allocator.free(prev);
    var curr = try allocator.alloc(i32, @intCast(n + 1));
    defer allocator.free(curr);

    for (prev) |*v| v.* = 0;

    var i: i32 = 1;
    while (i <= n) : (i += 1) {
        curr[0] = 0;
        var j: i32 = 1;
        while (j <= n) : (j += 1) {
            if (a[@intCast(i - 1)] == b[@intCast(j - 1)]) {
                curr[@intCast(j)] = prev[@intCast(j - 1)] + 1;
            } else {
                curr[@intCast(j)] = maxI32(prev[@intCast(j)], curr[@intCast(j - 1)]);
            }
        }
        const tmp = prev;
        prev = curr;
        curr = tmp;
    }

    return prev[@intCast(n)];
}

pub fn main(init: std.process.Init) !void {
    var it = init.minimal.args.iterate();
    _ = it.skip(); // program name

    const n_str = it.next() orelse {
        std.debug.print("Usage: lcs <n>\n", .{});
        std.process.exit(1);
    };

    const n = try std.fmt.parseInt(i32, n_str, 10);

    var buf: [32]u8 = undefined;
    var stdout = std.Io.File.stdout().writer(init.io, &buf);

    if (n <= 0) {
        try stdout.interface.print("{d}\n", .{@as(i32, 0)});
        try stdout.interface.flush();
        return;
    }

    const allocator = std.heap.page_allocator;
    const a = try allocator.alloc(u8, @intCast(n));
    defer allocator.free(a);
    const b = try allocator.alloc(u8, @intCast(n));
    defer allocator.free(b);

    fill(a, n, 0);
    fill(b, n, 0x9E3779B9);

    const result = try lcsLength(allocator, a, b, n);

    try stdout.interface.print("{d}\n", .{result});
    try stdout.interface.flush();
}
